src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -250,21 +262,36 @@ public void Reset()
 
         protected void ResetComponents()
         {
-            var frameHeaders = Interlocked.Exchange(ref _frameHeaders, null);
-            if (frameHeaders != null)
+            if(ServerOptions?.CacheHeadersAndStreams ?? false)
```

Comment:
Can just use `if (ServerOptions.CacheHeadersAndStreams)`.  Most Kestrel code assumes `ServerOptions` is not null.